SpokestackTray/Controllers/SpeechController.swift

```diff
@@ -0,0 +1,492 @@
+    @Published var stopOnErrorPublisher: CurrentValueSubject<Bool, Never> = CurrentValueSubject(false)
+    
+    @Published var didBeginSpeakingPublisher: CurrentValueSubject<Bool, Never> = CurrentValueSubject(false)
+    
+    @Published var didFinishSpeakingPublisher: CurrentValueSubject<Bool, Never> = CurrentValueSubject(false)
+    
+    @Published var didTimeoutPublisher: CurrentValueSubject<Bool, Never> = CurrentValueSubject(false)
+    
+    var configuration: TrayConfiguration!
+    
+    var isSilent: Bool = false
+    
+    var sayGreeting: Bool {
+        
+        set {
+            self.configuration.sayGreeting = newValue
+        }
+        get {
+            return self.configuration.sayGreeting
+        }
+    }
+    
+    var hasOnboarded: Bool {
+        set {
+            UserDefaults.standard.set(newValue, forKey: "TrayHasOnboarded")
+        }
+        get {
+            UserDefaults.standard.bool(forKey: "TrayHasOnboarded")
+        }
+    }
+    
+    var hasGreeted: Bool {
+        
+        set {
+            UserDefaults.standard.set(newValue, forKey: "TrayHasGreeted")
+        }
+        get {
+            UserDefaults.standard.bool(forKey: "TrayHasGreeted")
+        }
+    }
+
+    lazy var initializePublisher: AnyPublisher<(Bool, Downloads),TraySetupInitializeError>  = {
+
+        var errorMappedDownloadsPublisher: AnyPublisher<Downloads, TraySetupInitializeError> {
+            
+            return self.downloadsPublisher.mapError{ error -> TraySetupInitializeError in
+                
+                if let error = error as? TraySetupInitializeError {
+                    return error
+                }
+                
+                if let urlerror = error as? URLError {
+                    return TraySetupInitializeError.invalidModelDownloadStatus
+                }
+                
+                return TraySetupInitializeError.unknown
+            }
+            .eraseToAnyPublisher()
+        }
+        
+        return Publishers.CombineLatest(self.permissions.hasGrantedPermissions, errorMappedDownloadsPublisher).eraseToAnyPublisher()
+    }()
+    
+    // MARK: Private (properties)
+
+    private var tts: TextToSpeech?
+    
+    private var nlu: NLUTensorflow?
+    
+    private var canStart: Bool = true
+    
+    private var cancelleables: Set<AnyCancellable> = []
+    
+    lazy private var speechConfiguration: SpeechConfiguration = {
+        return SpeechConfiguration()
+    }()
+    
+    private var permissions: Permission = Permission()
+    
+    private var downloadsPublisher: AnyPublisher<Downloads, Error> {
+
+        let nluDownloadPublishers = Publishers.Zip3(downloadNLU, downloadNLUMetaData, downloadNLUVocab)
+        let wakeWordDownloadPublishers = Publishers.Zip3(downloadWakeWordFilter, downloadWakeWordDetect, downloadWakeWordEncode)
+        
+        return Publishers.Zip(nluDownloadPublishers, wakeWordDownloadPublishers).map{nluDownloads, wakeWordDownloads -> Downloads in
```

Comment:
```suggestion
        return Publishers.Zip(nluDownloadPublishers, wakeWordDownloadPublishers).map{(nluMeta, nluModel, nluVocab),(wwFilter, wwEncode, wwDetect) -> Downloads in
```

Reply:
Where you able to get this syntax to work? I'm getting compiler errors

Reply:
weird, it totally compiled for me that night, but I opened it again and it's failing

Reply:
you can get the same effect by restructuring using a let inside the block, but I guess I was wrong about tuple argument restructuring: https://forums.swift.org/t/pitch-tuple-destructuring-in-parameter-lists/2486/2

Reply:
@noelweichbrodt I couldn't get the above syntax to work. I had to add an additional mapping to get the tuple values to named parameters. Thoughts?
`        let nluDownloadPublishers = Publishers.Zip3(downloadNLU, downloadNLUMetaData, downloadNLUVocab)
        let wakeWordDownloadPublishers = Publishers.Zip3(downloadWakeWordFilter, downloadWakeWordDetect, downloadWakeWordEncode)
        
        let publishers = Publishers.Zip(nluDownloadPublishers, wakeWordDownloadPublishers).map{nluDownloads, wakeWordDownloads -> ((nluMeta: URL, nluModel: URL, nluVocab: URL),(wwFilter: URL, wwEncode: URL, wwDetect: URL)) in
            
            return ((nluMeta: nluDownloads.0, nluModel: nluDownloads.1, nluVocab: nluDownloads.2), (wwFilter: wakeWordDownloads.0, wwEncode: wakeWordDownloads.1, wwDetect: wakeWordDownloads.2))
        }
        
        return publishers.map{nluDownloads, wakeWordDownloads -> Downloads in
            
            return Downloads.init(nluMetaURL: nluDownloads.nluMeta,
                                  nluURL: nluDownloads.nluModel,
                                  nluVocabURL: nluDownloads.nluVocab,
                                  wakeWordFilter: wakeWordDownloads.wwFilter,
                                  wakeWordEncode: wakeWordDownloads.wwEncode,
                                  wakeWordDetect: wakeWordDownloads.wwDetect)
        }
        .receive(on: RunLoop.main)
        .eraseToAnyPublisher()`

Reply:
meh, the tuple is just used in one place, easy enough to see what it's capturing, so just disregard my initial suggestion please. swift should add tuple parameter destructuring!